routers/common/repo.go

```diff
@@ -52,55 +53,62 @@ func ServeData(ctx *context.Context, name string, size int64, reader io.Reader)
 
+	fileName := path.Base(filePath)
 	st := typesniffer.DetectContentType(buf)
+	isPlain := st.IsText() || ctx.FormBool("render")
+	mimeType := ""
+	cs := ""
 
-	mappedMimeType := ""
 	if setting.MimeTypeMap.Enabled {
-		fileExtension := strings.ToLower(filepath.Ext(name))
-		mappedMimeType = setting.MimeTypeMap.Map[fileExtension]
+		fileExtension := strings.ToLower(filepath.Ext(fileName))
+		mimeType = setting.MimeTypeMap.Map[fileExtension]
+	}
+
+	if mimeType == "" {
+		if st.IsBrowsableBinaryType() {
+			mimeType = st.GetMimeType()
+		} else if isPlain {
+			mimeType = "text/plain"
+		} else {
+			mimeType = typesniffer.ApplicationOctetStream
+		}
 	}
-	if st.IsText() || ctx.FormBool("render") {
-		cs, err := charset.DetectEncoding(buf)
+
+	if isPlain {
+		cs, err = charset.DetectEncoding(buf)
 		if err != nil {
-			log.Error("Detect raw file %s charset failed: %v, using by default utf-8", name, err)
+			log.Error("Detect raw file %s charset failed: %v, using by default utf-8", filePath, err)
 			cs = "utf-8"
 		}
-		if mappedMimeType == "" {
-			mappedMimeType = "text/plain"
-		}
-		ctx.Resp.Header().Set("Content-Type", mappedMimeType+"; charset="+strings.ToLower(cs))
+	}
+
+	if cs != "" {
+		ctx.Resp.Header().Set("Content-Type", mimeType+"; charset="+strings.ToLower(cs))
 	} else {
-		ctx.Resp.Header().Set("Access-Control-Expose-Headers", "Content-Disposition")
-		if mappedMimeType != "" {
-			ctx.Resp.Header().Set("Content-Type", mappedMimeType)
-		}
-		if (st.IsImage() || st.IsPDF()) && (setting.UI.SVG.Enabled || !st.IsSvgImage()) {
-			ctx.Resp.Header().Set("Content-Disposition", fmt.Sprintf(`inline; filename="%s"`, name))
-			if st.IsSvgImage() || st.IsPDF() {
-				ctx.Resp.Header().Set("Content-Security-Policy", "default-src 'none'; style-src 'unsafe-inline'; sandbox")
-				ctx.Resp.Header().Set("X-Content-Type-Options", "nosniff")
-				if st.IsSvgImage() {
-					ctx.Resp.Header().Set("Content-Type", typesniffer.SvgMimeType)
-				} else {
-					ctx.Resp.Header().Set("Content-Type", typesniffer.ApplicationOctetStream)
-				}
-			}
-		} else {
-			ctx.Resp.Header().Set("Content-Disposition", fmt.Sprintf(`attachment; filename="%s"`, name))
-		}
+		ctx.Resp.Header().Set("Content-Type", mimeType)
 	}
+	ctx.Resp.Header().Set("X-Content-Type-Options", "nosniff")
+
+	// serve types that can present a security risk with CSP
+	if st.IsSvgImage() {
+		ctx.Resp.Header().Set("Content-Security-Policy", "default-src 'none'; style-src 'unsafe-inline'; sandbox")
+	} else if st.IsPDF() {
+		// no sandbox attribute for pdf as it breaks rendering in at least safari. this
+		// should generally be safe as scripts inside PDF can not escape the PDF document
+		// see 	https://bugs.chromium.org/p/chromium/issues/detail?id=413851 for more discussion
+		ctx.Resp.Header().Set("Content-Security-Policy", "default-src 'none'; style-src 'unsafe-inline'")
+	}
+
+	ctx.Resp.Header().Set("Content-Disposition", `inline; filename*=UTF-8''`+url.PathEscape(fileName))
```

Comment:
just prompting to confirm this bit is written as intended? `filename*=UTF-8''`
(a comment explaining that syntax would be helpful)

Reply:
Yes, this is the weird syntax that https://datatracker.ietf.org/doc/html/rfc5987 imposes.

Reply:
added a comment